### PR TITLE
Automate testing of scap-security-guide

### PR DESCRIPTION
**Pull request by lilyeyes**
Automate testing of scap-security-guide: add '# oscap "eval" / "eval remote" / "remediate" / "eval after remediate"' into openQA

poo#93886 - [sle][security][sle15sp4]automate testing of scap-security-guide
poo#104943 - [sle][security][sle15sp4]automate testing of scap-security-guide: run '# oscap "eval" / "eval remote" / "remediate" / "eval after remediate"'

Note: the failure can be tracked by known product issue: bsc#1194676

- Related ticket:
  https://progress.opensuse.org/issues/104943
  https://progress.opensuse.org/issues/93886
- Needles: NA
- Verification run:
  x86_64: https://openqa.suse.de/tests/7983379
  s390x: https://openqa.suse.de/tests/7983467
  aarch64: https://openqa.suse.de/tests/7983465
  ppc64le: https://openqa.suse.de/tests/7983582 (wip, still has issue)

### CODE REVIEW

**Comment by jouyingbin** [comment]
note: POO is unable to access currently. https://progress.opensuse.org/issues/104943

File: tests/security/stig/oscap_security_guide_setup.pm [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    # Record the pkgs' version for reference
+    my $out = script_output("zypper se -s openscap-utils scap-security-guide");
+    record_info("Pkg_ver", "open scap security guide packages' version: $out");
+
+    # Set ds file
+    $self->set_ds_file();
+
+    # Check the ds file information for reference
+    my $f_ssg_sle_ds = $stigtest::f_ssg_sle_ds;
+    $out = script_output("oscap info $stigtest::f_ssg_sle_ds");
```
Comment (Amrysliu) [inline tests/security/stig/oscap_security_guide_setup.pm:31]: should use the parameter `$f_ssg_sle_ds`, it's been defined above.

Comment (lilyeyes) [reply tests/security/stig/oscap_security_guide_setup.pm:31]: Good pointing. I will revise.

**Comment by lilyeyes** [comment]
> note: POO is unable to access currently. https://progress.opensuse.org/issues/104943

It works now.

File: lib/stigtest.pm [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# Summary: Base module for STIG test cases
+# Maintainer: llzhao <llzhao@suse.com>
+
+package stigtest;
+
+use strict;
+use warnings;
+use testapi;
+use utils;
+
```
Comment (rfan1) [inline lib/stigtest.pm:15]: How about removing this line?

File: lib/stigtest.pm [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+# The file names of scap logs and reports
+our $f_stdout = 'stdout';
+our $f_stderr = 'stderr';
+our $f_report = 'report.html';
+
+# Set default value for 'scap-security-guide' ds file
+our $f_ssg_sle_ds = '/usr/share/xml/scap/ssg/content/ssg-sle12-ds.xml';
+
+# Profile ID
+our $profile_ID = 'xccdf_org.ssgproject.content_profile_stig';
```
Comment (rfan1) [inline lib/stigtest.pm:38]: How about adding a blank line here?

File: lib/stigtest.pm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    # for SLE15 the ds file is "ssg-sle15-ds.xml";
+    # for SLE12 the ds file is "ssg-sle12-ds.xml"
+    my $version = get_required_var('VERSION') =~ s/([0-9]+).*/$1/r;
+    $f_ssg_sle_ds = '/usr/share/xml/scap/ssg/content/ssg-sle' . "$version" . '-ds.xml';
+}
+
+sub upload_logs_reports
+{
+    # Upload logs & ouputs for reference
+    my $files = script_output('ls | grep "^ssg-sle.*.xml"');
```
Comment (rfan1) [inline lib/stigtest.pm:54]: @lilyeyes One minor question here, the files are located under /root directory, correct? I didn't see the abs path

File: tests/security/stig/oscap_xccdf_eval.pm [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+use warnings;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+
+    select_console 'root-console';
+
+    # Get ds file and profile ID, ds file
```
Comment (rfan1) [inline tests/security/stig/oscap_xccdf_eval.pm:19]: Suggest change 
# Get ds file and profile ID of ds file?

File: tests/security/stig/oscap_xccdf_eval_remote.pm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+use utils;
+use bootloader_setup qw(add_grub_cmdline_settings);
+use power_action_utils "power_action";
+
+sub run {
+    my ($self) = @_;
+
+    add_grub_cmdline_settings('ignore_loglevel', update_grub => 1);
+    power_action('reboot', textmode => 1);
+    $self->wait_boot(textmode => 1);
```
Comment (rfan1) [inline tests/security/stig/oscap_xccdf_eval_remote.pm:21]: @lilyeyes 
Kindly reminder, please conside ppc64le to handle reconnect_console issue

File: tests/security/stig/oscap_xccdf_eval_remote.pm [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    select_console 'root-console';
+
+    # Get ds file and profile ID, ds file
```
Comment (rfan1) [inline tests/security/stig/oscap_xccdf_eval_remote.pm:25]: Suggest changes:
Get ds file and profile ID of ds file

File: tests/security/stig/oscap_xccdf_eval_remote.pm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    my $profile_ID = $stigtest::profile_ID;
+    my $f_stdout = $stigtest::f_stdout;
+    my $f_stderr = $stigtest::f_stderr;
+    my $f_report = $stigtest::f_report;
+
+    # Verify detection mode with remote
+    my $ret = script_run("oscap xccdf eval --profile $profile_ID --oval-results --fetch-remote-resources --report $f_report $f_ssg_sle_ds > $f_stdout 2> $f_stderr", timeout => 3000);
+    record_info("Return=$ret", "# oscap xccdf eval --fetch-remote-resources --profile $profile_ID\" returns: $ret");
+    if ($ret == 137) {
+        $self->result("fail");
```
Comment (rfan1) [inline tests/security/stig/oscap_xccdf_eval_remote.pm:36]: Given that we have bug opened, how about add soft_fail tag here? or adding the bug number?

File: tests/security/stig/oscap_security_guide_setup.pm [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+use base 'stigtest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+
```
Comment (jouyingbin) [inline tests/security/stig/oscap_security_guide_setup.pm:16]: how about removing this blank line?

File: tests/security/stig/oscap_xccdf_eval.pm [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+use base 'stigtest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_eval.pm:16]: how about removing this blank line?

File: tests/security/stig/oscap_xccdf_remediate.pm [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+use base 'stigtest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_remediate.pm:16]: how about removing this blank line?

File: tests/security/stig/oscap_security_guide_setup.pm [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    select_console 'root-console';
+
+    # Install packages
+    zypper_call('in openscap-utils scap-security-guide', timeout => 180);
+
+    # Record the pkgs' version for reference
+    my $out = script_output("zypper se -s openscap-utils scap-security-guide");
+    record_info("Pkg_ver", "open scap security guide packages' version: $out");
```
Comment (jouyingbin) [inline tests/security/stig/oscap_security_guide_setup.pm:24]: is it a typo for **openscap** rather than open scap?

Comment (lilyeyes) [reply tests/security/stig/oscap_security_guide_setup.pm]: s/open scap/openscap/ but both are ok

File: tests/security/stig/oscap_xccdf_remediate.pm (hunk 11 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_remediate.pm]: done

File: tests/security/stig/oscap_xccdf_eval.pm (hunk 10 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval.pm]: done

File: tests/security/stig/oscap_security_guide_setup.pm (hunk 9 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_security_guide_setup.pm]: done

File: tests/security/stig/oscap_xccdf_eval_remote.pm (hunk 8 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval_remote.pm:37]: I keep "fail" and add bug number.

File: tests/security/stig/oscap_xccdf_eval_remote.pm (hunk 7 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval_remote.pm]: deleted ", ds file" (it is typo)

File: tests/security/stig/oscap_xccdf_eval_remote.pm (hunk 6 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval_remote.pm:21]: I tried 2 more times, test case can pass on s390x. so keep the code unmodified.

File: tests/security/stig/oscap_xccdf_eval.pm (hunk 5 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval.pm]: Deleted ", ds file" (typo)

File: lib/stigtest.pm (hunk 4 above)
Comment (lilyeyes) [reply lib/stigtest.pm:54]: It is under current working dir ($pwd).

File: lib/stigtest.pm (hunk 3 above)
Comment (lilyeyes) [reply lib/stigtest.pm:37]: done

File: lib/stigtest.pm (hunk 2 above)
Comment (lilyeyes) [reply lib/stigtest.pm]: done

File: tests/security/stig/oscap_security_guide_setup.pm (hunk 1 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_security_guide_setup.pm]: done

**Comment by jouyingbin** [comment]
@lilyeyes Would you provide the VR for SLE12 as you have the code about ssg-sle12-ds.xml ?

File: lib/stigtest.pm [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+package stigtest;
+
+use strict;
+use warnings;
+use testapi;
+use utils;
+use base 'opensusebasetest';
+
+our @EXPORT = qw(
+  $profile_ID
```
Comment (jouyingbin) [inline lib/stigtest.pm:18]: @lilyeyes  would you please list the items base on the code sequence?

File: tests/security/stig/oscap_xccdf_eval.pm [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    my $f_stdout = $stigtest::f_stdout;
+    my $f_stderr = $stigtest::f_stderr;
+    my $f_report = $stigtest::f_report;
+
+    # Verify detection mode
+    my $ret = script_run("oscap xccdf eval --profile $profile_ID --oval-results --report $f_report $f_ssg_sle_ds > $f_stdout 2> $f_stderr", timeout => 300);
+    if ($ret == 0) {
+        record_info('PASS');
+    } elsif ($ret == 1 || $ret == 2) {
+        record_info("errno=$ret", "# oscap xccdf eval --profile $profile_ID\" returns: $ret");
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_eval.pm:30]: just curious about it shows in openQA 
`# oscap xccdf eval --profile xccdf_org.ssgproject.content_profile_stig" returns: 2`
there is one **"** after stig, is it the correct behavior for your testing?

File: tests/security/stig/oscap_xccdf_eval_remote.pm [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+use power_action_utils "power_action";
+
+sub run {
+    my ($self) = @_;
+
+    add_grub_cmdline_settings('ignore_loglevel', update_grub => 1);
+    power_action('reboot', textmode => 1);
+    $self->wait_boot(textmode => 1);
+
+    select_console 'root-console';
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_eval_remote.pm:23]: why no put  **select_console 'root-console';** after **my ($self) = @_;** here ?

File: tests/security/stig/oscap_xccdf_eval_remote.pm [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+    my $f_stdout = $stigtest::f_stdout;
+    my $f_stderr = $stigtest::f_stderr;
+    my $f_report = $stigtest::f_report;
+
+    # Verify detection mode with remote
+    my $ret = script_run("oscap xccdf eval --profile $profile_ID --oval-results --fetch-remote-resources --report $f_report $f_ssg_sle_ds > $f_stdout 2> $f_stderr", timeout => 3000);
+    record_info("Return=$ret", "# oscap xccdf eval --fetch-remote-resources --profile $profile_ID\" returns: $ret");
+    if ($ret == 137) {
+        record_info('bsc#1194724');
+        $self->result("fail");
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_eval_remote.pm:37]: should be $self->result('fail');

**Comment by jouyingbin** [comment]
@lilyeyes Could you please explain there is **oscap_xccdf_eval#1** in the VR ?

File: tests/security/stig/oscap_xccdf_remediate.pm [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+sub run {
+    my ($self) = @_;
+    select_console 'root-console';
+
+    # Get ds file and profile ID, etc
```
Comment (jouyingbin) [inline tests/security/stig/oscap_xccdf_remediate.pm:18]: s/ etc / **etc.**

**Comment by jouyingbin** [comment]
@lilyeyes Not understand what is the status https://openqa.suse.de/tests/7983379#step/oscap_xccdf_remediate/27 in oscap_xccdf_remediate case, can you explain it? thanks.

**Comment by lilyeyes** [comment]
> @lilyeyes Not understand what is the status https://openqa.suse.de/tests/7983379#step/oscap_xccdf_remediate/27 in oscap_xccdf_remediate case, can you explain it? thanks.
After remediate the OS will pop up this message as some authentification check takes effect. It is by design.

**Comment by lilyeyes** [comment]
> @lilyeyes Could you please explain there is **oscap_xccdf_eval#1** in the VR ?

If the same "*pm" is executed in a job the next one will be marked as #1.

File: tests/security/stig/oscap_xccdf_eval_remote.pm (hunk 16 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval_remote.pm:37]: done

File: tests/security/stig/oscap_xccdf_eval_remote.pm (hunk 15 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval_remote.pm:23]: because "reboot"

File: tests/security/stig/oscap_xccdf_eval.pm (hunk 14 above)
Comment (lilyeyes) [reply tests/security/stig/oscap_xccdf_eval.pm:30]: changed to \"$profile_ID\"

File: lib/stigtest.pm (hunk 13 above)
Comment (lilyeyes) [reply lib/stigtest.pm:18]: I don't understand you question

**Review by rfan1 — APPROVED** [review]
LGTM!

File: lib/stigtest.pm [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  upload_logs_reports
+);
+
+# The file names of scap logs and reports
+our $f_stdout = 'stdout';
+our $f_stderr = 'stderr';
+our $f_report = 'report.html';
+
+# Set default value for 'scap-security-guide' ds file
+our $f_ssg_sle_ds = '/usr/share/xml/scap/ssg/content/ssg-sle12-ds.xml';
```
Comment (jouyingbin) [inline lib/stigtest.pm:34]: @lilyeyes Would you provide the VR for SLE12 as you have the code about ssg-sle12-ds.xml ?

Comment (lilyeyes) [reply lib/stigtest.pm:34]: I have discussed with Marcus I will only support sle15. IMO qem team might be like to add it to their job group.

File: lib/stigtest.pm (hunk 13 above)
Comment (jouyingbin) [reply lib/stigtest.pm:18]: I mean the list sequence as following variables you defined in this pm file.

Comment (lilyeyes) [reply lib/stigtest.pm:18]: Oh, got it.
I think it is not necessary or maybe not good.
I just list the parameters together which are similar, it is readable somehow. such as list them together: $f_stdout
  $f_stderr   $f_report

**Review by Amrysliu — APPROVED** [review]

File: lib/stigtest.pm (hunk 18 above)
Comment (jouyingbin) [reply lib/stigtest.pm:34]: got it :)

**Review by jouyingbin — APPROVED** [review]
After discussing with @lilyeyes offline, everything LGTM now. thanks.

**Pull request merged**